src/backend/distributed/executor/executor.c

```diff
@@ -1431,6 +1615,9 @@ TransactionStateMachine(WorkerSession *session)
 			{
 				if (execution->isTransaction)
 				{
+					/* if we're expanding the nodes in a transaction, use 2PC */
+					Activate2PCIfModifyingTransactionExpandsToNewNode(session);
```

Comment:
I think this is probably the perfect place.

Reply:
Ah, this actually doesn't work well with the following:

```SQL
select count(*) from users_table where user_id = 1;
select count(*) from users_table where user_id = 2;
insert into users_table (user_id) VALUES (1);
insert into users_table (user_id) VALUES (2);
```

Because the transaction is not `REMOTE_TRANS_INVALID ` but `REMOTE_TRANS_SENT_BEGIN`

Reply:
Well, the main issue that we cannot a nice place for this is the `ForceMaxQueryParallelization` because with that the placements are assigned before the connection establishment is done whereas in the regular execution path, the placements are assigned after the connection establishment is done. Otherwise, calling the function on `StartPlacementExecutionOnSession()` would be logical and fix the issue at all.

I really cannot come up with a place other than `MULTI_CONNECTION_CONNECTED`. I can undo the flag check and move back to there. Otherwise, I had to end-up with calling the function in multiple places, which I don't like.

Or, do you have any other suggestions?

Reply:
I much prefer calling it in several places over calling it in the main loop. Where could we call it to support `ForceMaxQueryParallelization`?

Reply:
makes sense, please see my latest commit(s). I think the second point where I added the call fixed both the edge cases because of `ForceMaxQueryParallelization ` and selects opening remote transaction blocks. I added the tests as well